mycodo/utils/method.py

```diff
@@ -1,115 +1,94 @@
-                    P0[0] - x]
-    # Find roots of the polynomial to determine the parameter t
-    roots = np.roots(coefficients)
-    # Find the root which is between 0 and 1, and is also real
-    correct_root = None
-    for root in roots:
-        if np.isreal(root) and 0 <= root <= 1:
-            correct_root = root
-    # Check a valid root was found
-    if correct_root is None:
-        print('Error, no valid root found. Are you sure your Bezier curve '
-              'represents a valid function when projected into the xy-plane?')
-        return 0
-    param_t = correct_root
-    # From the value for the t parameter, find the corresponding y-value
-    # using the formula for cubic Bezier curves
-    y = (1-param_t)**3*P0[1] + 3*(1-param_t)**2*param_t*P1[1] + 3*(1-param_t)*param_t**2*P2[1] + param_t**3*P3[1]
-    assert np.isreal(y)
-    # Typecast y from np.complex128 to float64
-    y = y.real
-    return y
-
 
-def calculate_method_setpoint(method_id, table, this_controller, Method, MethodData, logger):
+    def __init__(self, method_data, method_type, logger=None):
+        """
+        Initializes the method class
+        :param method_data: data queried from method_data table
+        :param method_type: method type from method table
+        :param logger: The logger to use
+        :return: 0 (success) or 1 (error) and a setpoint value
+        """
+        self.logger = logger
+
+        self.method_type = method_type
+
+        self.method_data = method_data
+
+        self.method_data_all = self.method_data.filter(MethodData.output_id.is_(None)).all()
+        self.method_data_first = self.method_data.filter(MethodData.output_id.is_(None)).first()
+        self.method_data_repeat = self.method_data.filter(MethodData.duration_sec == 0).first()
+
+    def calculate_setpoint(self, now, method_start_time=None):
+        """
+        Returns the value for the setpoint for a given point in time and elapsed duration
+        :param now: point in time to calculate the value for
+        :param method_start_time: when this method started. Must accept datetime and strings
+        :return: float value of the setpoint; True if method finished, otherwise False
+        """
+        return None, False
+
+    def should_restart(self, now, method_start_time=None):
+        """
+        If a method has signalled to be finished, this method is asked if the controller should restart
+        the method processing from the beginning.
+        :param now: point in time to calculate the value for
+        :param method_start_time: when this method started. Must accept datetime and strings
+        :return: True if the method wants to be restarted, otherwise False
+        """
+        return False
+
+    def get_plot(self, max_points_x=None):
+        """
+        Called to calculate values to render a plot in the frontend.
+        :param max_points_x:
+        :return:  2d array with x and y values.
+        """
+        return []
```

Comment:
This is called by the frontend to generate plot data. The code from the frontend was replaced by calls to the calculate_set_point implementations to get rid of duplcated calculation code.